datafusion-examples/examples/async_udf.rs

```diff
@@ -0,0 +1,405 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use arrow_schema::{Field, Schema, SchemaRef};
+use async_trait::async_trait;
+use datafusion::{
+    arrow::{
+        array::{ArrayRef, Float32Array, Float64Array},
+        datatypes::DataType,
+        record_batch::RecordBatch,
+    },
+    execution::{
+        context::{QueryPlanner, SessionState},
+        runtime_env::RuntimeEnv,
+        TaskContext,
+    },
+    logical_expr::Volatility,
+    physical_expr::PhysicalSortExpr,
+    physical_plan::{
+        stream::RecordBatchStreamAdapter, DisplayFormatType, Distribution, ExecutionPlan,
+        Partitioning, SendableRecordBatchStream, Statistics,
+    },
+    physical_planner::{DefaultPhysicalPlanner, ExtensionPlanner, PhysicalPlanner},
+};
+
+use datafusion::prelude::*;
+use datafusion::{error::Result, physical_plan::functions::make_scalar_function};
+use datafusion_common::{
+    cast::{as_float32_array, as_float64_array},
+    tree_node::{Transformed, TreeNode},
+    DFSchemaRef, DataFusionError,
+};
+use datafusion_expr::{
+    expr::ScalarUDF, Extension, LogicalPlan, Subquery, UserDefinedLogicalNode,
+    UserDefinedLogicalNodeCore,
+};
+use datafusion_optimizer::{optimize_children, OptimizerConfig, OptimizerRule};
+use std::{
+    any::Any,
+    fmt::{self, Debug},
+    sync::Arc,
+};
+
+use futures::{FutureExt, StreamExt};
+
+// create local execution context with an in-memory table
```

Comment:
I think it would help in this file somewhere to give the larger context -- namely that this example is showing how to use the datafusion extension points (user defined nodes, and rewriters) to implement an arbitrary user defined asynchronous table function.

Reply:
I agree. I updated the documentation. I am just beginning to understand how DataFusion works and hope the wording is correct.